lib/actions/apiV2.js

```diff
@@ -701,12 +742,183 @@ export const findPatternsForRoute = (params) =>
+  ),
+  from: {
+    ...leg.from,
+    stopCode: leg.from.stop?.code,
+    stopId: leg.from.stop?.gtfsId
+  },
+  route: leg.route?.shortName,
+  routeColor: leg.route?.color,
+  routeId: leg.route?.id,
+  routeLongName: leg.route?.longName,
+  routeShortName: leg.route?.shortName,
+  routeTextColor: leg.route?.textColor,
+  to: {
+    ...leg.to,
+    stopCode: leg.to.stop?.code,
+    stopId: leg.to.stop?.gtfsId
+  },
+  tripHeadsign: leg.trip?.tripHeadsign,
+  tripId: leg.trip?.gtfsId
+})
+
+export function routingQuery(searchId = null, updateSearchInReducer) {
+  return function (dispatch, getState) {
+    const state = getState()
+    const { config, currentQuery } = state.otp
+    const persistenceMode = getPersistenceMode(config.persistence)
+    const activeItinerary = getActiveItinerary(state)
+
+    const isNewSearch = !searchId
+    if (isNewSearch) searchId = randId()
+    // Don't permit a routing query if the query is invalid
+    if (!queryIsValid(state)) {
+      console.warn('Query is invalid. Aborting routing query', currentQuery)
+      return
+    }
+
+    const { date, departArrive, routingType, time } = currentQuery
+    const arriveBy = departArrive === 'ARRIVE'
+
+    const { activeModes, modeSettings } = getActivatedModesFromQueryParams(
+      state.router.location.search,
+      state.otp.config.modes.modeButtons,
+      [
+        ...defaultModeSettings,
+        ...(state.otp.config.modes.modeSettingDefinitions || [])
+      ],
+      state.otp.config.modes.initialState
+    )
+    // Get the raw query param strings to save for the rider's search history
+    const urlSearchParams = new URLSearchParams(state.router.location.search)
+    const rawModeButtonQP = urlSearchParams.get('modeButtons')
+    const rawModeSettingQP = urlSearchParams.get('modeSettings')
+
+    const combinations = generateCombinations({
+      arriveBy,
+      date,
+      from: currentQuery.from,
+      modes: activeModes,
+      modeSettings,
+      numItineraries: config?.modes?.numItineraries || 7,
+      time,
+      to: currentQuery.to
+    })
+
+    dispatch(
+      routingRequest({
+        activeItinerary,
+        pending: combinations.length,
+        routingType,
+        searchId,
+        updateSearchInReducer: false
+      })
+    )
+
+    dispatch(setItineraryView(ItineraryView.LIST))
+
+    combinations.forEach((combo) => {
+      const query = generateOtp2Query(combo)
+      dispatch(
+        createGraphQLQueryAction(
+          query.query,
+          query.variables,
+          (response) => {
+            const dispatchedRoutingResponse = routingResponse(response)
+            // If tracking is enabled, store locations and search after successful
+            // search is completed.
+            if (
+              persistenceMode.isLocalStorage &&
+              state.user?.localUser?.storeTripHistory
+            ) {
+              const { from, to } = currentQuery
+              if (!isStoredPlace(from)) {
```

Comment:
Moving them outside the loop increases the function's complexity too high. Maybe I can simplify it somewhere else. I don't want to make a new dispatched action for this since that would require a ton of argument passing that would be messy.

Reply:
I couldn't find a nice way of doing this. Moving these evaluations outside of the loop increases the function's complexity too much. I think this really should just all be refactored in a future PR.